### DISCUSSION TITLE
relevance of boundaryTypeMap

### DISCUSSION THREAD

**Discussion by turbulentphil** [Nek5000/nekRS#571] [Q&A] [answered]
Dear Neks,

I am coming from Nek5000, trying to get started with NekRS, and wondering what the use of the parameters `boundaryTypeMap` in the _.par_ file is.

Here is what I can observe from example conj_ht (as an example):
1) The mesh files _fluid.box_ and _solid.box_ already specify the boundary condition types.
2) The _.par_ file specifies `boundaryTypeMap = v, O, W` and `boundaryTypeMap = t, O, I` for the velocity and temperature field, respectively. Such statements are new to me.
3) The _.usr_ file specifies in `usrdat2` some integer values for `boundaryID` and `boundaryIDt` based on the boundary condition type from the mesh files.
4) Interestingly, `usrdat2` seems to associate the values {1, 2, 3} for boundary types {v, O, W} (from the velocity mesh), which are the indices of these letters in `boundaryTypeMap`. (Q0: Is this a coincidence?)

My questions are the following:
Q1) What is the use of `boundaryTypeMap`? Is it maybe just adding flexibility for people that didn't use Nek5000/`genbox` to generate the mesh?
Q2) Is there any reason to use the new `boundaryTypeMap` parameters in addition to boundary types specified in the mesh files using `genbox`?
Q3) Assuming the relation between `usrdat2` and `boundaryTypeMap` (as in point 4), could one also skip these steps given a definition of the mesh as done in _fluid.box_ and _solid.box_?

I would appreciate any help.

Best,
Philipp

**Comment by yslan** [comment, marked as answer]
Hi Philipp,

`boundaryTypeMap` is another way to specify the boundary conditions. This is also introduced in recent Nek5000 since this [version](https://github.com/Nek5000/Nek5000/pull/787), but they work slightly differently.

- If your `.re2` already has boundary conditions and you are happy with it, you don't need to set anything and NekRS will read Nek5000's CBC arrays for BC. Meaning, you can still modify the CBC in `userdat2` in Nek5000 way.
- If you want, you can also set boundary conditions via `boundaryID`. Basically, you name each boundary an integer (1-base, start from 1, cannot skip numbers), then you can assign boundary conditions in `.par` via `boundaryTypeMap`. For example, `boundaryTypeMap = v, O, W` indicating `v` (inflow) for `boundaryID=1`, `O` (outflow) for `boundaryID=2` and `W` (wall) for `boundaryID=3`
- `boundaryIDt` is only used when the case is a conjugate heat transfer case, or `nelgt>nelgv`. Then, the temperature boundary will use `boundaryIDt` for t-mesh.

This should answer your Q0 and Q2
Q1. Yes, you can consider it as a new feature. Especially for meshes from gmsh2nek that stores bc id in boundaryID.
Q3. No, you still need to tag BC in box files (or somehow recover it in userdat2) otherwise the code won't be able to distinguish the faces. 

Hope this helps,
Yu-Hsiang

**Reply by turbulentphil** [reply]
Hi Yu-Hsiang,

So I understand that (1) if the `.re2` file already contains the `cbc` arrays (and specifies the boundary conditions sufficiently), then I don't necessarily need to deal with `boundaryID` or `boundaryTypeMap`. Otherwise, (2) the integers assigned in `usrdat2` are the indices from `boundaryTypeMap` and can be used alternatively to specify the boundary conditions.
Thanks a lot!

Best,
Philipp
